index.bs

```diff
@@ -1074,6 +1085,221 @@ spec: webidl
+    document defines the following <a>errors</a>, algorithms, and <a>extension
+    commands</a> for the [[WebDriver]] specification.
+  </p>
+
+  The <dfn>no such request</dfn> <a>error</a> has a [=response/status=] of 404
+  and an a JSON error code of `"no such request"`.
+
+  <div algorithm="detect-pending-request">
+    <p>
+      To <dfn lt="detect a pending request">detect a pending request</dfn>
+      for a JSON object |permissionDesc| and string |type|, the UA must perform
+      the following steps.
+    </p>
+    <ol>
+      <li>Let |rootDesc| be the object |permissionDesc| refers to, <a>converted
+      to an IDL value</a> of type {{PermissionDescriptor}}. If this throws an
+      exception, return <a>error</a> with <a>error code</a> <a>invalid
+      argument</a>.</li>
+      <li>Let |typedDescriptor| be the object |permissionDesc| refers to,
+      <a>converted to an IDL value</a> of
+      <code>|rootDesc|.{{PermissionDescriptor/name}}</code>'s <a>permission
+      descriptor type</a>. If this throws an exception, return <a>error</a>
+      with <a>error code</a> <a>invalid argument</a>.</li>
+      <li>If |type| is "`use`":
+        <ol>
+          <li>*EachUseRequest*: For each <a>request permission to use</a>
+            operation that is awaiting user input:
+            <ol>
+              <li>Let |pendingDescriptor| be the {{PermissionDescriptor}}
+                associated with the request.</li>
+              <li>[=map/For each=] |key| → |value| of |pendingDescriptor|:
+                <ol>
+                  <li>If |typedDescriptor|[|key|] does not equal |value|, continue
+                  to the next iteration of the loop labeled
+                  *EachUseRequest*.</li>
+                </ol>
+              </li>
+              <li>Return true.</li>
+            </ol>
+          </li>
+        </ol>
+      </li>
+      <li>Else:
+        <ol>
+          <li class="assertion">Assert: |type| is "`choose`".
+          <li>*EachChooseRequest*: For each <a>prompt the user to choose</a>
+          operation that is awaiting user input:
+            <ol>
+              <li>Let |pendingDescriptor| be the {{PermissionDescriptor}}
+                associated with the request.</li>
+              <li>[=map/For each=] |key| → |value| of |pendingDescriptor|:
+                <ol>
+                  <li>If |typedDescriptor|[|key|] does not equal |value|, continue
+                  to the next iteration of the loop labeled
+                  *EachChooseRequest*.</li>
+                </ol>
+              </li>
+              <li>Return true.</li>
+            </ol>
+          </li>
+        </ol>
+      </li>
+      <li>Return false.
+    </ol>
+  </div>
+
+  <section>
+    <h3 id="grant-command">
+      Grant
+    </h3>
+    <table>
+      <tbody>
+        <tr>
+          <th>HTTP Method</th>
+          <th><a lt="extension command prefix">Prefix</a></th>
```

Comment:
What will the resulting URLs look like? https://w3c.github.io/webdriver/webdriver-spec.html#protocol-extensions talks about "vendor-specific needs" and "Commands that are specific to a user agent are called extension commands", which isn't the case here.

Should we perhaps just be defining new commands in the style of other commands in the WebDriver spec with a "Path Template"? If the resulting URLs are the same it doesn't matter, but I suspect not, and it'd be a bit silly if you could tell from the URL where something was specified.

Reply:
According to the above issue, I think what is here is good. The extension command prefix is "permissions" and the extension command name is "set". So the entire command will look like session/{:sessionId}/permissions/set which is in line with the example.

Reply:
Is there any point having /get and /set? It seems more in line with the rest of WebDriver to use GET and POST verbs towards the same URL template.

Reply:
Sorry for the late response, I agree with @burg. It would be more simple to have the command be POST session/{:sessionId}/permissions
Also I think it might make sense to have a get permissions command here, but that can come later.

Reply:
I’ve filed https://github.com/w3c/webdriver/issues/1142 to fix the slant towards vendor-specific extensions. It is the intention to make extension commands a useful entry-point for other specifications to define commands that could be useful to enable testing.

The current prose is slanted towards vendor-specific extensions because that was the first use case we had for them, and this is the first specification that tries to hook into WebDriver. If the current method hooking into WebDriver are sub-optimal we can fix that.

Reply:
> Is there any point having /get and /set? It seems more in line with the rest
> of WebDriver to use GET and POST verbs towards the same URL template.

This proposal does not include a /get endpoint, but I think your point about idiom remains.

The WebDriver specification requires that all extension commands be specified with a "prefix" and a "name", which are combined to form the extension command's URI template. When I authored this patch, I could only satisfy that requirement by specifying some string, so I chose "Set." In light of recent recommendations about the "session ID", I think we *could* circumvent this by specifying the "prefix" as "session/{session id}" and the "name" as "permissions". However, this doesn't really jive with the intent of the specification:

> separating extension commands from other commands in order to avoid potential
> resource conflicts with other implementations.

So I think we'll need to revise the WebDriver spec if we want a nicer URL. I've proposed that here https://github.com/w3c/webdriver/issues/1170